packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -292,73 +292,85 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
     final VoidCallback? previousTrainHoppingListenerRemover = _trainHoppingListenerRemover;
     _trainHoppingListenerRemover = null;
 
-    if (nextRoute is TransitionRoute<dynamic> && canTransitionTo(nextRoute) && nextRoute.canTransitionFrom(this)) {
-      final Animation<double>? current = _secondaryAnimation.parent;
-      if (current != null) {
-        final Animation<double> currentTrain = (current is TrainHoppingAnimation ? current.currentTrain : current)!;
-        final Animation<double> nextTrain = nextRoute._animation!;
-        if (
-          currentTrain.value == nextTrain.value ||
-          nextTrain.status == AnimationStatus.completed ||
-          nextTrain.status == AnimationStatus.dismissed
-        ) {
-          _setSecondaryAnimation(nextTrain, nextRoute.completed);
-        } else {
-          // Two trains animate at different values. We have to do train hopping.
-          // There are three possibilities of train hopping:
-          //  1. We hop on the nextTrain when two trains meet in the middle using
-          //     TrainHoppingAnimation.
-          //  2. There is no chance to hop on nextTrain because two trains never
-          //     cross each other. We have to directly set the animation to
-          //     nextTrain once the nextTrain stops animating.
-          //  3. A new _updateSecondaryAnimation is called before train hopping
-          //     finishes. We leave a listener remover for the next call to
-          //     properly clean up the existing train hopping.
-          TrainHoppingAnimation? newAnimation;
-          void _jumpOnAnimationEnd(AnimationStatus status) {
-            switch (status) {
-              case AnimationStatus.completed:
-              case AnimationStatus.dismissed:
-                // The nextTrain has stopped animating without train hopping.
-                // Directly sets the secondary animation and disposes the
-                // TrainHoppingAnimation.
-                _setSecondaryAnimation(nextTrain, nextRoute.completed);
+    if (nextRoute is TransitionRoute<dynamic>) {
+      if (canTransitionTo(nextRoute) && nextRoute.canTransitionFrom(this)) {
```

Comment:
I could reduce a whole bunch of diffs here if I used an `else if` below the existing logic, rather than adding a whole new layer of logical nesting. Happy to do so if y'all would prefer. Nothing here until line 359 has been changed at all.